### PR TITLE
OHM-871 BAHMNI Mapping Functions

### PR DESCRIPTION
**Pull request by MattyJ007**
Dubnium is the LTS and testing latest node is useful.

The previous version 0.10 being tested is no longer relevant

OHM-871

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -1,39 +1,23 @@
 {
-  "author": "Dan VerWeire <dverweire@gmail.com>",
+  "author": "Jembi health Systems NPC",
   "name": "object-mapper",
-  "description": "Copy properties from one object to another.",
-  "version": "5.0.0",
+  "description": "Map fields from one object to another user a user defined schema",
+  "version": "1.0.0",
```
Comment (BMartinos) [inline package.json]: Lets maybe start this at `0.1.0` as an initial version and once we are happy with everything we can bump to an official 1 release.

File: README.md [hunk 2]
```diff
@@ -9,8 +9,10 @@ on a separate `Object`, which defines how the properties should be mapped.
 
 ## Installation
 
-```shell
-$ npm install --save object-mapper
+In your projects `package.json` file include the following line in the dependencies field:
+
+```json
+"object-mapper": "jembi/node-object-mapper#master"
```
Comment (BMartinos) [inline README.md]: Do we need to supply the hash for master? would it not default to master in any case?

File: README.md [hunk 3]
```diff
@@ -68,30 +71,44 @@ var map = {
   "foo": [
     {
       key: "foo",
-      transform: function (value) { 
-        return value + "_foo";
+      transform:
+        {
```
Comment (BMartinos) [inline README.md]: Maybe just for consistency lets move the brackets after the colon instead of a new line

File: src/transform/codes.js [hunk 4]
```diff
@@ -0,0 +1,19 @@
...
+      case 'M':
+        return 'male'
+      case 'F':
+        return 'female'
+      case 'O':
+        return 'other'
+      default:
+        return 'unknown'
+    }
+  }
```
Comment (BMartinos) [inline src/transform/codes.js]: As mentioned to you in a chat, it would be great if we could get rid of these hard coded cases which might become problematic when adding more cases for other input values. The idea we could work on is to have these cases built up based on the user supplied parameters set in the `input-mapping.json`. This will make it more flexible to the implementing user and less maintenance on the developers.

File: src/transform/primitives.js [hunk 5]
```diff
@@ -0,0 +1,14 @@
...
+// primitive data types
+
+exports.primitives = {
+  booleanFlip: function(val) {
+    if (typeof val === 'boolean') {
+      return !val
+    }
+
+    return null
+  }
```
Comment (BMartinos) [inline src/transform/primitives.js:27]: Maybe as a future add, but we add add another function here to do a `booleanConvert` function. This will be useful in the case where a value is a string but meant to be Boolean. This will take in the value supplied and make a boolean out of it. The below example should return a boolean of `false`
E.g:
```
{
  booleanValue: "false"
}
```

Comment (BMartinos) [reply src/transform/primitives.js:27]: This will become important when we can start chaining function together. So if the value is a string of false, but we need to populate it as a Boolean true we can chain the `booleanConvert` first and then the `booleanFlip` functions

File: src/transform/tests.js [hunk 6]
```diff
@@ -0,0 +1,45 @@
...
+
+exports.tests = {
+  'test-foo': function(value) {
+    return value + '-foo'
+  },
+  'test-destination': function(val, _src, dst) {
+    dst.manual = val.a + val.b
+  },
+  'test-empty-array': function(sourceValue) {
+    var i
```
Comment (BMartinos) [inline src/transform/tests.js]: Not sure if this was exiting or not, but I think we can remove this declaration here and do it in the for loop instead.

**Review by BMartinos — CHANGES_REQUESTED** [review]
Thanks @MattyJ007 looking good so far, just a few minor comments

File: src/transform/tests.js (hunk 6 above)
Comment (MattyJ007) [reply src/transform/tests.js]: Lol, that was from the original tests - went straight from eyes to fingers by-passing brain

File: README.md [hunk 7]
```diff
@@ -210,60 +209,60 @@ mapping defined by **mapObject**
...
 This function is also exported directly from `require('object-mapper')` (ie: `var merge = require('object-mapper');`)
 
-* **sourceObject** is the object FROM which properties will be copied.
-* **destinationObject** [OPTIONAL] is the object TO which properties will be copied.
-* **mapObject** is the object which defines how properties are copied from
-**sourceObject** to **destinationObject**
+- **sourceObject** is the object FROM which properties will be copied.
+- **destinationObject** [OPTIONAL] is the object TO which properties will be copied.
+- **mapObject** is the object which defines how properties are copied from
+  **sourceObject** to **destinationObject**
```
Comment (BMartinos) [inline README.md]: Should this not be on the same line?

**Review by BMartinos — APPROVED** [review]
LGTM :+1:

File: test/transformationsTest.js [hunk 8]
```diff
@@ -0,0 +1,291 @@
...
+        null,
+        null,
+        functionParameters
+      )
+
+      t.equal(result, expectedOutput)
+      t.end()
+    })
+
+    t.test('should convert unhandled code to null', t => {
```
Comment (BMartinos) [inline test/transformationsTest.js]: Im not sure this test title reflects the actual test case. It seems to be testing the logic same as the previous test

Comment (MattyJ007) [reply test/transformationsTest.js]: :| Whoops

File: test/transformationsTest.js [hunk 9]
```diff
@@ -0,0 +1,291 @@
...
+        t.equal(result, expectedOutput)
+        t.end()
+      }
+    )
+
+    t.test(
+      'should convert unhandled codes to null if no default specified',
+      t => {
+        const inputCode = 7
+        const expectedOutput = undefined
```
Comment (BMartinos) [inline test/transformationsTest.js:73]: Should this be `undefined` or `null`?

Comment (MattyJ007) [reply test/transformationsTest.js:73]: Undefined :| 
I'm not sure if that will be an issue or not for object-mapper... I would assume not?

File: test/transformationsTest.js [hunk 10]
```diff
@@ -0,0 +1,291 @@
...
+        const result = transforms.booleanFlip(booleanInput)
+
+        t.equal(result, expectedOutput)
+        t.end()
+      }
+    )
+
+    t.test('should return null if input is a falsy', t => {
+      const booleanInput = 0
+      const expectedOutput = null
```
Comment (BMartinos) [inline test/transformationsTest.js:208]: Would this not be returning `false` instead of `null`?
https://github.com/jembi/node-object-mapper/pull/1/files#diff-0dec99e818a2242edae67563675d4d06R25

Comment (BMartinos) [reply test/transformationsTest.js:208]: oh wait nvm, thats a different function call...

Comment (MattyJ007) [reply test/transformationsTest.js:208]: Nope - this is testing the booleanFlip function which won't accept truthys and falsys.

booleanify will change 0 to false

File: test/transformationsTest.js (hunk 9 above)
Comment (BMartinos) [reply test/transformationsTest.js:73]: Im not to sure but dont think so, just need to update the test title to state `undefined` insted of `null` :)

**Review by BMartinos — APPROVED** [review]
